### PR TITLE
add UpdatableAdapter.

### PR DESCRIPTION
**Pull request by shy1st**
add UpdatableAdapter.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@shy1st remove useless commits:

image

**Comment by hsluoyz** [comment]
@shy1st plz use another PR for adding license header.

**Comment by hsluoyz** [comment]
@shy1st any update?

**Comment by hsluoyz** [comment]
@shy1st plz fix CI error like:

image

File: src/main/java/org/casbin/jcasbin/util/StringPool.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ *  Author: DreamLu 卢春梦 (596392912@qq.com)
+ */
+package org.casbin.jcasbin.util;
+
+/**
+ * 静态 String 池
+ *
+ * @author L.cm
+ */
+public interface StringPool {
```
Comment (hsluoyz) [inline src/main/java/org/casbin/jcasbin/util/StringPool.java:24]: I think we don't need this file, just using "," as comma would be OK.

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: src/main/java/org/casbin/jcasbin/util/StringPool.java (hunk 1 above)
Comment (shy1st) [reply src/main/java/org/casbin/jcasbin/util/StringPool.java]: Thanks for your advice, I will deal with it.

File: src/main/java/org/casbin/jcasbin/model/Policy.java [hunk 2]
```diff
@@ -14,13 +14,18 @@
 
 package org.casbin.jcasbin.model;
 
+import org.apache.commons.lang3.StringUtils;
```
Comment (hsluoyz) [inline src/main/java/org/casbin/jcasbin/model/Policy.java:17]: What is this for?

File: pom.xml [hunk 3]
```diff
@@ -117,6 +117,11 @@
             <artifactId>commons-io</artifactId>
             <version>2.6</version>
         </dependency>
+        <dependency>
```
Comment (hsluoyz) [inline pom.xml:120]: What is this for?

File: src/main/java/org/casbin/jcasbin/effect/DefaultEffector.java [hunk 4]
```diff
@@ -30,23 +30,23 @@ public DefaultEffector() {
     @Override
     public boolean mergeEffects(String expr, Effect[] effects, float[] results) {
         boolean result;
-        if (expr.equals("some(where (p_eft == allow))")) {
+        if ("some(where (p_eft == allow))".equals(expr)) {
```
Comment (hsluoyz) [inline src/main/java/org/casbin/jcasbin/effect/DefaultEffector.java:33]: why change this?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: src/main/java/org/casbin/jcasbin/model/Assertion.java [hunk 5]
```diff
@@ -29,6 +31,7 @@
     public String value;
     public String[] tokens;
     public List<List<String>> policy;
+    public Map<String, Integer> policyMap;
```
Comment (hsluoyz) [inline src/main/java/org/casbin/jcasbin/model/Assertion.java:34]: Remove it.

File: src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java [hunk 6]
```diff
@@ -28,91 +31,109 @@
...
-            buildIncrementalRoleLinks(Model.PolicyOperations.POLICY_ADD, ptype, rules);
-        }
-
-        if (watcher != null && autoNotifyWatcher) {
-            if (watcher instanceof WatcherEx) {
-                ((WatcherEx) watcher).updateForAddPolicy(rule.toArray(new String[0]));
-            } else {
-                watcher.update();
+            model.addPolicy(sec, ptype, rule);
+            if (sec.equals(Primitive.GROUP)) {
```
Comment (hsluoyz) [inline src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java:59]: Why do we add `Primitive.GROUP`?

**Comment by hsluoyz** [comment]
Nice work!

**Comment by shy1st** [comment]
> Nice work!

Thanks.

**Pull request merged**